README
draw.io project sources
-----------------------

draw.io comprises three parts:

1) The static web content under the war directory.
2) The open, save, embed, proxy and iconfinder servlets as one servlet war. These are located under src/com/mxgraph/online/.
3) The image export servlet. This is located under etc/imageexport.

Prior to invoking any builds, search war/index.html for "CUSTOM_PARAMETERS". The image export URL is the most important variable to change (EXPORT_URL). 
Without that the export to image will use our server, which somewhat defeats the point of hosting it. The servlet is mapped to "export" and the default name 
of the war is "Imageexport" so "/Imageexport/export" is the value if you deploy that war as-is and the servlet engine is on the root of the domain.

/etc/build/build.xml is the ant build file for the main war. Invoking "war" will generate build/draw.war.

/etc/imagexport/build.xml is the ant build file for the image export war. Invoking "war" will generate /etc/imageexport/imageexport.war.

Depoly these wars on a servlet engine and navigate to the location of the first war, you will get the draw.io application.

ChangeLog
25-OCT-2012: 1.6.9.0

- Adds Google Drive install scope
- Updates libraries used for Google Drive integration

19-OCT-2012: 1.6.8.1

- Fixes handling of file positioning in Drive

16-OCT-2012: 1.6.8.0

- Fixes using multiple accounts with Google Drive

24-SEP-2012: 1.6.7.4

- Changes diagram.ly to draw.io in dialogs
- Minor formatting fixes

21-SEP-2012: 1.6.7.3

- Add session tidying cron job for GAE

14-SEP-2012: 1.6.7.2

- Removes Facebook footer in IE6
- Adds layout options via grapheditor example

13-SEP-2012: 1.6.7.1

- Fixes use of custom handlers with closure compiler

31-AUG-2012: 1.6.7.0

- Uses mxGraph 1.10.3.1
- Fixes background color for SVG export in quirks mode
- Enables Google image picker on Mac/iPad
- Adds Github menu item in help menu

29-AUG-2012: 1.6.1.9

- Fixes Github deployment Ant task
- Fixes German Facebook button (changes to English)
- Fixes touch-based panning for embedded diagrams
- Uses mxGraph 1.10.3.0

22-AUG-2012: 1.6.1.8

- Adds Facebook like button

21-AUG-2012: 1.6.1.7

- Adds compression for embedded diagrams

02-AUG-2012: 1.6.1.6

- Adds fixes for null user objects in link handling

02-AUG-2012: 1.6.1.5

- Adds transparency option for PNG image, SVG export

01-AUG-2012: 1.6.1.4

- Fixes for filepicker.io

31-JUL-2012: 1.6.1.3

- Uses mxClient.js, mxgraph-core.jar 1.10.2.1
- Adds social, analytics URL parameters
- Adds persistent stats for embedding
- Adds save for filepicker.io

28-JUL-2012: 1.6.1.2

- Fixes embedding of international characters and newlines
- Changes image export URL to http://shr.diagramly.com:8080/ImageExport/export
- Changes lastModified header creation in EmbedServlet
- Fixes infinite recursion in Embed.js for convertValueToString

27-JUL-2012: 1.6.1.1

- Adds support for hyperlinks

23-JUL-2012: 1.6.1.0

- Uses mxGraph 1.10.2.0
- Adds embed feature
- Fixes errors if tinyMCE not available
- Fixes syntax of dash patterns in shapes

30-JUN-2012: 1.6.0.2

- Renames Init.js to Devel.js
- Adds link to "Did you know..." topic in help menu
- Removes previous diagramly link from help menu
- Adds url URL parameter handling in client
- Adds proxy servlet for loading files from URLs

22-JUN-2012: 1.6.0.1

- Fixes minor bugs
- Uses mxGraph 1.10.1.3

19-JUN-2012: 1.6.0.0

- Adds real-time collaboration (experimental)
- Adds handling for App not installed error in Google Drive
- Uses mxGraph 1.10.1.2

06-JUN-2012: 1.5.0.7

- Uses mxGraph 1.10.1.1
- Adds workaround for focus repaint bug in IE9/10
- Adds fixed connection points for custom shapes
- Adds closure compiler pre-deploy step
- Puts all JS in a single minified file

25-MAY-2012: 1.5.0.6

- Uses static mxGraph 1.10.0.6
- Fixes "Save as" in Google Drive
- Fixes page view, default pageScale
- Adds new items in file menu
- Adds options in file edit dialog
- Adds import action
- Adds extra handle for custom shapes

17-MAY-2012: 1.5.0.5

- Removes .mxe default file extension
- Changes mime type to application/mxe

16-MAY-2012: 1.5.0.4

- Uses current protocol to load picker API
- Adds CSS3 transitions for hover styles
- Simplifies tinyMCE toolbar, uses main toolbar (non-IE only)
- Adds "Use as default edge" action

14-MAY-2012: 1.5.0.3

- Adds border in export dialog
- Adds resources for Google Drive integration
- Fixes use of optional flash for saving in savedialog

05-MAY-2012: 1.5.0.2

- Handles requests for non-existent script versions
- Handles DeadlineExceededException

03-MAY-2012: 1.5.0.1

- Adds XML export option
- Fixes unicode character encoding in files
- Adds status information for Google Drive operations
- Adds spinner during Google Drive operations
- Adds reconnect for lost session after save
- Fixes illegal state exceptions in servlet

02-MAY-2012: 1.5.0.0

- Adds working Google Drive integration

01-MAY-2012: 1.4.0.7

- Adds initial Google Drive integration

29-APR-2012: 1.4.0.6

- Uses mxGraph 1.10.0.2
- Adds BPMN shapes, markers and custom edges
- Fixes Downloadify integration

19-APR-2012: 1.4.0.5

- Adds editFile action from grapheditor
- Adds fromText action and ParseDialog
- Adds forum link in help menu
- Adds more shapes sidebar entry
- Adds iconfinder library

29-MAR-2012: 1.4.0.2

- Adds UML library
- Adds experimental group selection
- Adds persistent UI settings
- Adds templates, file, new... menu
- Adds HTML labels with editing
- Adds aa URL parameter to disable anti aliasing
- Adds PDF export via backend with limited HTML support
- Adds context menu in sidebar
- Various fixes from bug reports, disabled Flash

16-FEB-2012: 1.4.0.0

- First release of new codebase

24-JAN-2012: 1.3.2.2

- Fixes for brix

13-JAN-2012: 1.3.2.0

- Enables .vdx importing for GAE

25-DEC-2011: 1.3.1.2

- Avoids passing null version parameter onto primary client server

23-DEC-2011: 1.3.1.1

- Adds AWS icons
- Updates core to 1.8.0.6

15-DEC-2011: 1.3.1.0

- Adds local SVG generation
- Adds client servlet for cached mxclient delivery

17-NOV-2011: 1.3.0.1

- mxgraph.mockuop.form_elements stencil set name typo corrected

11-NOV-2011: 1.3.0.0

- Addition stencil based shapes added

14-JUL-2011: 1.2.0.1

- Collaboration integrated

11-JUL-2011: 1.2.0.0

- Deployed to GAE by default

24-JUN-2011: 1.1.1.0

- Sprited toolbar icons
- Local open for file API enabled browsers

26-MAY-2011: 1.1.0.0

- Switches most content to static serving
- New iPad interface

12-MAY-2011: 1.0.11.0

- Adds simplified default stylesheet for new diagrams
- Adds UI for transparency, user-defined colors and images

02-MAY-2011: 1.0.10.1

- Additional static gzipping of resources
- Adds lazy loading of images in collapsed palettes

05-APR-2011: 1.0.10.0

- Fixes string escaping parse errors on open
- Add show/parse XML options

17-MAR-2011: 1.0.9.0

- Adds duplicate context menu item, ctrl-D
- Adds optional navigation context menu
- Adds optional navigation icons
- Adds support for cursor keys
- Adds guides

05-MAR-2011: 1.0.8.0

- Adds connect preview
- Adds splash screens
- Adds import of VDX files


war/js/diagramly/Sidebar.js
(function()
{
	// Overrides gear image URL
	Sidebar.prototype.gearImage = GRAPH_IMAGE_PATH + '/clipart/Gear_128x128.png';
	
	/**
	 * Toggle palette.
	 */
	Sidebar.prototype.togglePalettes = function(prefix, ids)
	{
		for (var i = 0; i < ids.length; i++)
		{
			this.togglePalette(prefix + ids[i]);
		}
	};

	/**
	 * Toggle palette.
	 */
	Sidebar.prototype.togglePalette = function(id)
	{
		var elts = this.palettes[id];
		
		if (elts != null)
		{
			var vis = (elts[0].style.display == 'none') ? 'block' : 'none';
			
			for (var i = 0; i < elts.length; i++)
			{
				elts[i].style.display = vis;
			}
		}
	};

	// Replaces the sidebar
	Sidebar.prototype.init = function()
	{
		var imgDir = GRAPH_IMAGE_PATH;
		var dir = STENCIL_PATH;
		var signs = this.signs;
		var mockups = this.mockups;
		var ee = this.ee;
		var pids = this.pids;

		this.addGeneralPalette(true);
		this.addIconfinder();
		this.addUmlPalette(false);

		this.addBpmnPalette(dir, false);
		this.addStencilPalette('flowchart', 'Flowchart', dir + '/flowchart.xml',
			';fillColor=#ffffff;strokeColor=#000000;strokeWidth=2');
		this.addStencilPalette('basic', mxResources.get('basic'), dir + '/basic.xml',
			';fillColor=#ffffff;strokeColor=#000000;strokeWidth=2');
		this.addStencilPalette('arrows', mxResources.get('arrows'), dir + '/arrows.xml',
			';fillColor=#ffffff;strokeColor=#000000;strokeWidth=2');
		this.addImagePalette('computer', 'Clipart / Computer', imgDir
				+ '/lib/clip_art/computers/', '_128x128.png', [ 'Antivirus',
				'Data_Filtering', 'Database', 'Database_Add', 'Database_Minus',
				'Database_Move_Stack', 'Database_Remove', 'Fujitsu_Tablet',
				'Harddrive', 'IBM_Tablet', 'iMac', 'iPad', 'Laptop', 'MacBook',
				'Mainframe', 'Monitor', 'Monitor_Tower',
				'Monitor_Tower_Behind', 'Netbook', 'Network', 'Network_2',
				'Printer', 'Printer_Commercial', 'Secure_System', 'Server',
				'Server_Rack', 'Server_Rack_Empty', 'Server_Rack_Partial',
				'Server_Tower', 'Software', 'Stylus', 'Touch', 'USB_Hub',
				'Virtual_Application', 'Virtual_Machine', 'Virus',
				'Workstation' ]);
		this.addImagePalette('finance', 'Clipart / Finance', imgDir
				+ '/lib/clip_art/finance/', '_128x128.png', [ 'Arrow_Down',
				'Arrow_Up', 'Coins', 'Credit_Card', 'Dollar', 'Graph',
				'Pie_Chart', 'Piggy_Bank', 'Safe', 'Shopping_Cart',
				'Stock_Down', 'Stock_Up' ]);
		this.addImagePalette('clipart', 'Clipart / Various', imgDir
				+ '/lib/clip_art/general/', '_128x128.png', [ 'Battery_0',
				'Battery_100', 'Battery_50', 'Battery_75', 'Battery_allstates',
				'Bluetooth', 'Earth_globe', 'Empty_Folder', 'Full_Folder',
				'Gear', 'Keys', 'Lock', 'Mouse_Pointer', 'Plug', 'Ships_Wheel',
				'Star', 'Tire' ]);
		this.addImagePalette('networking', 'Clipart / Networking', imgDir
				+ '/lib/clip_art/networking/', '_128x128.png', [ 'Bridge',
				'Certificate', 'Certificate_Off', 'Cloud', 'Cloud_Computer',
				'Cloud_Computer_Private', 'Cloud_Rack', 'Cloud_Rack_Private',
				'Cloud_Server', 'Cloud_Server_Private', 'Cloud_Storage',
				'Concentrator', 'Email', 'Firewall_02', 'Firewall',
				'Firewall-page1', 'Ip_Camera', 'Modem',
				'power_distribution_unit', 'Print_Server',
				'Print_Server_Wireless', 'Repeater', 'Router', 'Router_Icon',
				'Switch', 'UPS', 'Wireless_Router', 'Wireless_Router_N' ]);
		this.addImagePalette('people', 'Clipart / People', imgDir
				+ '/lib/clip_art/people/', '_128x128.png', [ 'Suit_Man',
				'Suit_Man_Black', 'Suit_Man_Blue', 'Suit_Man_Green',
				'Suit_Man_Green_Black', 'Suit_Woman', 'Suit_Woman_Black',
				'Suit_Woman_Blue', 'Suit_Woman_Green',
				'Suit_Woman_Green_Black', 'Construction_Worker_Man',
				'Construction_Worker_Man_Black', 'Construction_Worker_Woman',
				'Construction_Worker_Woman_Black', 'Doctor_Man',
				'Doctor_Man_Black', 'Doctor_Woman', 'Doctor_Woman_Black',
				'Farmer_Man', 'Farmer_Man_Black', 'Farmer_Woman',
				'Farmer_Woman_Black', 'Nurse_Man', 'Nurse_Man_Black',
				'Nurse_Man_Green', 'Nurse_Man_Red', 'Nurse_Woman',
				'Nurse_Woman_Black', 'Nurse_Woman_Green', 'Nurse_Woman_Red',
				'Military_Officer', 'Military_Officer_Black',
				'Military_Officer_Woman', 'Military_Officer_Woman_Black',
				'Pilot_Man', 'Pilot_Man_Black', 'Pilot_Woman',
				'Pilot_Woman_Black', 'Scientist_Man', 'Scientist_Man_Black',
				'Scientist_Woman', 'Scientist_Woman_Black', 'Security_Man',
				'Security_Man_Black', 'Security_Woman', 'Security_Woman_Black',
				'Soldier', 'Soldier_Black', 'Tech_Man', 'Tech_Man_Black',
				'Telesales_Man', 'Telesales_Man_Black', 'Telesales_Woman',
				'Telesales_Woman_Black', 'Waiter', 'Waiter_Black',
				'Waiter_Woman', 'Waiter_Woman_Black', 'Worker_Black',
				'Worker_Man', 'Worker_Woman', 'Worker_Woman_Black' ]);
		this.addImagePalette('telco', 'Clipart / Telecommunication', imgDir
				+ '/lib/clip_art/telecommunication/', '_128x128.png', [
				'BlackBerry', 'Cellphone', 'HTC_smartphone', 'iPhone',
				'Palm_Treo', 'Signal_tower_off', 'Signal_tower_on' ]);

		for (var i = 0; i < signs.length; i++)
		{
			this.addStencilPalette('signs' + signs[i], 'Signs / ' + signs[i],
				dir + '/signs/' + signs[i].toLowerCase() + '.xml',
				';fillColor=#000000;strokeColor=none');
		}
		
		for (var i = 0; i < mockups.length; i++)
		{
			this.addStencilPalette('ui' + mockups[i], 'Mockup / ' + mockups[i],
				dir + '/mockup/' + mockups[i].toLowerCase().replace(/ /g, '_') + '.xml');
		}
		
		for (var i = 0; i < ee.length; i++)
		{
			this.addStencilPalette('electrical' + ee[i], 'Electrical / ' + ee[i],
				dir + '/electrical/' + ee[i].toLowerCase().replace(/ /g, '_') + '.xml',
				';fillColor=white;strokeColor=black');
		}

		// Adds AWS stencils
		this.addStencilPalette('awsCompute', 'AWS / Compute', dir + '/aws/compute.xml', ';fillColor=#FF9800;strokeColor=none');
		this.addStencilPalette('awsContentDelivery', 'AWS / Content Delivery', dir + '/aws/content_delivery.xml', ';fillColor=#1EA4DD;strokeColor=none');
		this.addStencilPalette('awsDatabase', 'AWS / Database', dir + '/aws/database.xml', ';fillColor=#6F2D6E;strokeColor=none');
		this.addStencilPalette('awsDeploymentManagement', 'AWS / Deployment Management', dir + '/aws/deployment_management.xml', ';fillColor=#296934;strokeColor=none');
		this.addStencilPalette('awsGroups', 'AWS / Groups', dir + '/aws/groups.xml');
		this.addStencilPalette('awsMessaging', 'AWS / Messaging', dir + '/aws/messaging.xml', ';fillColor=#B8B58A;strokeColor=none');
		this.addStencilPalette('awsMisc', 'AWS / Misc', dir + '/aws/misc.xml', ';fillColor=#F7981F;strokeColor=none');
		this.addStencilPalette('awsNetworking', 'AWS / Networking', dir + '/aws/networking.xml', ';fillColor=#262261;strokeColor=none');
		this.addStencilPalette('awsNonServiceSpecific', 'AWS / Non Service Specific', dir + '/aws/non_service_specific.xml', ';fillColor=#C5C7C9;strokeColor=none');
		this.addStencilPalette('awsOnDemandWorkforce', 'AWS / On Demand Workforce', dir + '/aws/on_demand_workforce.xml', ';fillColor=#C5C7C9;strokeColor=none');
		this.addStencilPalette('awsStorage', 'AWS / Storage', dir + '/aws/storage.xml', ';fillColor=#146EB4;strokeColor=none');

		for (var i = 0; i < pids.length; i++)
		{
			this.addStencilPalette('pid' + pids[i], 'P&ID / ' + pids[i],
				dir + '/pid/' + pids[i].toLowerCase().replace(' ', '_') + '.xml',
				';fillColor=#ffffff;strokeColor=#000000');
		}
		
		this.addStencilPalette('leanMapping', 'Lean Mapping', dir + '/lean_mapping.xml',
			';fillColor=#ffffff;strokeColor=#000000;strokeWidth=2');
		this.addMoreShapes();
	};
	
	// Adds the more shapes entry
	Sidebar.prototype.addMoreShapes = function()
	{
		var elt = this.createTitle(mxResources.get('moreShapes') + '...');
		this.container.appendChild(elt);
		
		// Block handling as link in IE
		mxEvent.addListener(elt, 'click', function(evt)
		{
			mxEvent.consume(evt);
		});
		
		this.moreShapes = elt;
 	};
 	
	// Adds the iconfinder library
	Sidebar.prototype.addIconfinder = function()
	{
		// TODO: Fix delayed typing, occasional error in library creation in
		// quirks mode
		var elt = this.createTitle(mxResources.get('images'));
		this.container.appendChild(elt);
		
		var div = document.createElement('div');
		div.className = 'geSidebar';
	    div.style.display = 'none';
		div.style.overflow = 'hidden';
		div.style.width = '100%';
		div.style.padding = '0px';
		
		var inner = document.createElement('div');
		inner.className = 'geTitle';
		inner.style.backgroundColor = 'transparent';
		inner.style.borderColor = 'transparent';
		inner.style.padding = '4px';
		inner.style.textOverflow = 'clip';
		inner.style.cursor = 'default';
		
		if (!mxClient.IS_VML)
		{
			inner.style.paddingRight = '20px';
		}

		var searchResource = mxResources.get('search');
		
		var input = document.createElement('input');
		input.setAttribute('type', 'text');
		input.value = searchResource;
		input.style.border = 'solid 1px #d5d5d5';
		input.style.width = '100%';
		input.style.backgroundImage = 'url(' + IMAGE_PATH + '/clear.gif)';
		input.style.backgroundRepeat = 'no-repeat';
		input.style.backgroundPosition = '100% 50%';
		input.style.paddingRight = '14px';
		inner.appendChild(input);

		var cross = document.createElement('div');
		cross.setAttribute('title', mxResources.get('reset'));
		cross.style.position = 'relative';
		cross.style.left = '-16px';
		cross.style.width = '12px';
		cross.style.height = '14px';
		cross.style.cursor = 'pointer';

		// Workaround for inline-block not supported in IE
		cross.style.display = (mxClient.IS_VML) ? 'inline' : 'inline-block';
		cross.style.top = ((mxClient.IS_VML) ? 0 : 3) + 'px';
		
		// Needed to block event transparency in IE
		cross.style.background = 'url(' + IMAGE_PATH + '/transparent.gif)';
		
		var find;

		mxEvent.addListener(cross, 'click', function()
		{
			input.value = '';
			find();
			input.focus();
		});
		
		inner.appendChild(cross);
		div.appendChild(inner);

		var center = document.createElement('center');
		var button = mxUtils.button(searchResource, function()
		{
			find();
		});
		button.setAttribute('disabled', 'true');
		// Workaround for inherited line-height in quirks mode
		button.style.lineHeight = 'normal';
		center.style.paddingTop = '4px';
		center.style.marginBottom = '12px';
		
		center.appendChild(button);
		div.appendChild(center);
		
		var searchTerm = '';
		var modified = false;
		var active = false;
		var complete = false;
		var page = 0;
		var count = 25;
		
		function clearDiv()
		{
			var child = div.firstChild;
			
			while (child != null)
			{
				var next = child.nextSibling;
				
				if (child != inner && child != center)
				{
					child.parentNode.removeChild(child);
				}
				
				child = next;
			}
		};
		
		find = mxUtils.bind(this, function(callback)
		{
			if (input.value != '' || (!modified && input.value == searchResource))
			{
				if (button.getAttribute('disabled') != 'true')
				{
					if (center.parentNode != null)
					{
						if (searchTerm != input.value)
						{
							clearDiv();
							searchTerm = input.value;
							complete = false;
							page = 0;
						}
						
						if (!active)
						{
							button.innerHTML = mxResources.get('loading') + '...';
							active = true;
							mxUtils.get(ICONFINDER_PATH + '?q=' + encodeURIComponent(searchTerm) + '&c=' + count +
									'&p=' + page, mxUtils.bind(this, function(req)
							{
								active = false;
								page++;
								center.parentNode.removeChild(center);
								var icons = req.getXml().getElementsByTagName('icon');
								
								for (var i = 0; i < icons.length; i++)
								{
									var size = null;
									var url = null;
									var child = icons[i].firstChild;
									
									while (child != null && (size == null || url == null))
									{
										if (child.nodeName == 'size')
										{
											size = parseInt(mxUtils.getTextContent(child));
										}
										else if (child.nodeName == 'image')
										{
											url = mxUtils.getTextContent(child);
										}
		
										child = child.nextSibling;
									}
									
									if (size != null && url != null)
									{
										div.appendChild(this.createVertexTemplate('shape=image;image=' + url, size, size, ''));
									}
								}
								
								if (icons.length < count)
								{
									button.setAttribute('disabled', 'true');
									button.innerHTML = mxResources.get('noMoreResults');
									complete = true;
								}
								else
								{
									button.innerHTML = mxResources.get('moreResults');
								}
								
								if (icons.length == 0 && page == 1)
								{
									var err = document.createElement('div');
									err.className = 'geTitle';
									err.style.backgroundColor = 'transparent';
									err.style.borderColor = 'transparent';
									err.style.padding = '4px';
									err.style.textAlign = 'center';
									err.style.cursor = 'default';
									
									mxUtils.write(err, mxResources.get('noResultsFor', [searchTerm]));
									div.appendChild(err);
								}
								
								div.appendChild(center);
							}));
						}
					}
				}
			}
			else
			{
				clearDiv();
				searchTerm = '';
				button.innerHTML = searchResource;
				button.setAttribute('disabled', 'true');
			}
		});
		
		mxEvent.addListener(input, 'keydown', mxUtils.bind(this, function(evt)
		{
			if (evt.keyCode == 13 /* Enter */)
			{
				find();
			}
		}));
		
		mxEvent.addListener(input, 'keyup', mxUtils.bind(this, function(evt)
		{
			modified = true;
			
			if (input.value == '' || (!modified && input.value == searchResource))
			{
				button.setAttribute('disabled', 'true');
			}
			else if (input.value != searchTerm)
			{
				button.removeAttribute('disabled');
				button.innerHTML = searchResource;
			}
			else if (!active)
			{
				if (complete)
				{
					button.setAttribute('disabled', 'true');
					button.innerHTML = mxResources.get('noMoreResults');
				}
				else
				{
					button.removeAttribute('disabled');
					button.innerHTML = mxResources.get('moreResults');
				}
			}
		}));
		
		mxEvent.addListener(input, 'focus', mxUtils.bind(this, function(evt)
		{
			if (input.value == searchResource && !modified)
			{
				input.value = '';
			}
		}));
		
		mxEvent.addListener(input, 'blur', mxUtils.bind(this, function(evt)
		{
			if (input.value == '')
			{
				input.value = searchResource;
				modified = false;
			}
		}));
	    
	    // Workaround for blocked text selection in Editor
	    mxEvent.addListener(input, 'mousedown', function(evt)
	    {
	    	if (evt.stopPropagation)
	    	{
	    		evt.stopPropagation();
	    	}
	    	
	    	evt.cancelBubble = true;
	    });
	    
	    // Workaround for blocked text selection in Editor
	    mxEvent.addListener(input, 'selectstart', function(evt)
	    {
	    	if (evt.stopPropagation)
	    	{
	    		evt.stopPropagation();
	    	}
	    	
	    	evt.cancelBubble = true;
	    });
	    
		this.addFoldingHandler(elt, div, function()
		{
			// not lazy
		}, false);
	    
		var outer = document.createElement('div');
	    outer.appendChild(div);
	    this.container.appendChild(outer);
		
	    // Keeps references to the DOM nodes
    	this.palettes['images'] = [elt, outer];
 	};
})();